v-next/hardhat/src/internal/cli/init/template.ts

```diff
@@ -0,0 +1,83 @@
+import path from "node:path";
+
+import {
+  exists,
+  getAllFilesMatching,
+  isDirectory,
+  readdir,
+  readJsonFile,
+} from "@ignored/hardhat-vnext-utils/fs";
+import {
+  findClosestPackageRoot,
+  PackageJsonNotFoundError,
+  type PackageJson,
+} from "@ignored/hardhat-vnext-utils/package";
+
+export interface Template {
+  name: string;
+  packageJson: PackageJson;
+  path: string;
+  files: string[];
+}
+
+/**
+ * getTemplates returns the list of available templates. It retrieves them from
+ * the "templates" folder in the package root.
+ *
+ * @returns The list of available templates.
+ */
+export async function getTemplates(): Promise<Template[]> {
+  const packageRoot = await findClosestPackageRoot(import.meta.url);
+  const pathToTemplates = path.join(packageRoot, "templates");
+
+  if (!(await exists(pathToTemplates))) {
+    return [];
+  }
+
+  const pathsToTemplates = await readdir(pathToTemplates);
+
+  const templates = await Promise.all(
+    pathsToTemplates.map(async (name) => {
+      const pathToTemplate = path.join(pathToTemplates, name);
+      const pathToPackageJson = path.join(pathToTemplate, "package.json");
+
+      if (!(await isDirectory(pathToTemplate))) {
+        return;
+      }
+
+      // Validate that the the template has a package.json file
+      if (!(await exists(pathToPackageJson))) {
+        // eslint-disable-next-line no-restricted-syntax -- this is a valid usecase for the PackageJsonNotFoundError
+        throw new PackageJsonNotFoundError(pathToPackageJson);
+      }
```

Comment:
I don't think the eslint error should be ignored. In Hardhat we intentionally only throw `HardhatError`s, which are more helpful for a user.

But I'd say that more than a validation, this is checking an invariant that should always hold.

I would replace this with something like

```ts
assertHardhatInvarian(await exists(pathToPackageJson), `package.json for template ${name} is missing`);
```

The reason is that if that is `false`, it's an important bug that we need to fix.

Reply:
That's an excellent suggestion. We should definitely catch this before the code reaches our users. I really like the invariant solution - I added it now.